tenant/configloader/config.go

```diff
@@ -0,0 +1,22 @@
+package configloader
+
+import (
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/options/provideropts"
+)
+
+// configuration for the tenantLoader implementation (configloader)
+type Configuration struct {
+	TenantsFile string // path to the tenant's config file
+	Rules       []*RuleConfig
+}
+
+// structure that is loaded from the tenant's config file
```

Comment:
Oh wait a minute, so, once you have your provider configured, this adds a new dimension of tenants within the provider? This is very keycloak esque and doesn't translate well to other providers

Reply:
Rules are what match an incoming request to a tenant/provider. The ConfigLoader configures all the providers and keeps them in memory. Then in a middleware (`pkg/middleware/tenantLoader.go`), the request is matched to a tenant and then that tenant is injeted into the request context. The rules are there to do the matching. We execute rules one by one untill we find a match.
A rule (`type RuleConfig struct`) has following structure.

```
type RuleSource string

const (
   SourceHost        RuleSource = "host"
   SourcePath        RuleSource = "path"
   SourceQueryParams RuleSource = "query"
   SourceHeader      RuleSource = "header"
)

type RuleConfig struct {
  Source       RuleSource // which part of the HTTP request contains the tenant id
  Expr         string     // the regex expression to match and extract tenant id from the source
  CaptureGroup int        // capture group or sub-match that is actually the tenant id
  QueryParam   string     // query parameter in case source is 'query'
  Header       string     // header key in case source is 'header'
}
```

I think these rules are pretty generic and are not just related to keycloak. We just have to have tenant-id somewhere in the incoming request. That location is specified the `Source` field in a rule. Right now, host, path, header, and query are supported but we can expand to include cookies as well.